### ISSUE TITLE
Syntax proposal question: `value()` or `value.get()`

### ISSUE THREAD

**Issue by mweststrate** [mobxjs/mobx#120]
When creating an observable scalar, mobservable 1.0 will return a setter/ getter function. Like:

```javascript
const cityName = observable("Amsterdam");

// get and print
autorun(() => console.log(cityName());

// update value
cityName("Rotterdam");
```

In practice you probably won't use this that much, because most observables are part of an object anyway. But I was wondering, would it be preferable to make it more explicit that this is an observable by changing the syntax to:

```javascript
const cityName = observable("Amsterdam");

// get and print
autorun(() => console.log(cityName.get());

// update value
cityName.set("Rotterdam");
```

What do you think?

**Comment by mweststrate** [comment]
@kmalakoff Did you think about this one in your api simplification proposal?

I like the second example for it's explicitness and the fact that it saves a closure per scalar. The first one is nicer for academic purposes in the sense that it makes clear that you should think of values as expressions that yield different values over time.

A combination of the two would also be possible btw: `cityName()` + `cityName.set()`

**Comment by dslmeinte** [comment]
Hmmm, though one...

In any case, I'd pick only one and not provide the sugar.

I slightly lean towards the 2nd, more explicit version for pragmatic reasons (more readable and consistent with typical pre-existing mental models), although aesthetically I like the first one better.

**Comment by luisherranz** [comment]
My vote goes for the second one.

It's easier to know what's going on with
`cityName.get()` + `cityName.set()` than with
`cityName()` + `cityName.set()`.

**Comment by Kosta-Github** [comment]
same for me: option 2 ... same reasoning as from @luisherranz

**Comment by mweststrate** [comment]
Option 2 it will be!

**Comment by kmalakoff** [comment]
@mweststrate sorry for the delay. I was thinking we could deprecate this API and only use extend on objects or the equivalent class syntax. 

A few reasons:

1) reduce API surface area to keep the concepts simple
2) make the API analogous across ES5, ESNext, and TypeScript 
3) I haven't personally yet come a across a need for individual observables

I'd really like to see someone make a strong case for keeping them. Who out there is using them? Why? Would extend or the class analogous syntaxes give an equivalent result?

**Comment by mweststrate** [comment]
Hi Kevin,

You suggested this earlier and I think you are actually right. I use it mainly for testing and documentation, but both are not good reasons to keep it. I wan't make the change now for not having to many changes. But actually that might be a bad idea, we are bumping a major anyway after all.

Maybe we should also include a few other tweak while we are at it ;-) (most as proposed in #74):

# 1. Deprecate `observable(scalar)`
As sugested, deprecate the `observable` _function_ (not the decorator) for all scalar values including functions, null and undefined. It also makes `expr` more relevant: an unanymous `@computed` value

# 1b. Break down `observable`?
I'm not sure about this one, but we could then remove `(@)observable` completely and break it down into `(@)observableObject`, `(@)observableArray` and `(@)observableMap`. I don't really like it, except that this weird construction is no longer needed: `@observable dict = mobservable.map({ data })` which is a strangely more cumbersome than `@observable obj = {}`. However using `@observableMap dict = {}` would be incorrect typewise; right hand side is an object but it would become a Map. Not sure whether for example typescript support such situations.

# 2. introduced the `@computed` decorator
.. which is exactly the same as the `@observable` decorator on getter functions, so that these two concepts are explicitly different in the api although they both create something observable:

```javascript
class Person {
  @observable firstName;
  @observable lastName;
  @observable get fullName() { return this.firstName + this.lastName}
}

// would become:
class Person {
  @observable firstName;
  @observable lastName;
  @computed get fullName() { return this.firstName + this.lastName}
}
```

# 3. introduce modifiers on decorators
So than one can write:

```
class Person {
  @observable({ flat: true }) tags = [];
  // instead of
  @observable tags = asFlat([ ]);
}
```
The modifiers should continue to exist in there current form as well, to be able to use a modifier deep inside an object? Not sure whether anybody uses that feature

# 4. Remove `untracked`
It is an undocumented feature, I think hardly anyone is using it. A `peek` function would be a cleaner alternative if really needed.

**Comment by kmalakoff** [comment]
There's a lot to unpack here! 

Some of them are quite breaking changes so I would decide if any of them should just be removed from the docs, have deprecation warnings, and keep running until adequate time has passed. Alternatively, like you say you can treat it like a major release and people who need time to migrate wait for the upgrade until they are ready.

On each issue above:

# 1. Deprecate `observable(scalar)`

If I understand, that is what is thread is about so I'm all for this

# 1b. Break down observable?

Interesting one. I am not really familiar with TypeScript, but I might choose the solution that is most intuitive to a TypeScript developer and consistent with TypeScript since you can point to their rationale. 

If this isn't a good suggestion, maybe put it out to the community in a separate thread (personally, I have no issues with specialized names per type since there are only a few but I do like a smaller set of concepts / keywords)

# 2. introduced the @computed decorator

I find it difficult to understand when `get` is needed or not. I'm not sure if / when it is needed. If computed helps explain the `get` conceptually, it sounds great.

# 3. introduce modifiers on decorators

I'm not sure about this one. It is tied to the auto-nested-wrapping of observables and what the default behavior should be. As you remember, I'm confused by this and seem to want to turn it off, but others seem to expect it. I would break it out into another thread for examples and discussion.

# 4. Remove untracked

I use this a lot to break cycles. With 2.0, these problems should go away. If there is still a chance for cycles (eg. it actually still does something in 2.0), maybe just put a deprecation warning and remove in a subsequent release?  If it is now unnecessary, sure, pull it out since it is a simple search and replace to remove.


In summary:

1. Yes
1b. I'm on the fence (but probably more for the custom keywords unless there is a good way to be consistent with TypeScript). Discussion needed
2. Yes (but can `get` be removed for consistency with ES5?)
3. Not fully satisfied with the current and possible solutions. Discussion needed
4. Yes (if definitely not needed anymore)

**Comment by mweststrate** [comment]
* 1a. I'll deprecate it and mark it for deletion in 2.1
* 1b. I think it is fine to keep the current api. Didn't get complaints about it. When in doubt, less api changes is probably better
* 2. I think it's a good idea, i'll introduce `@computed` and mark `@observable get ...()` for deletion in 2.1
* 3. this is an incremental change so can be done in any later version. Some people find it confusing indeed, on the other hand, usually it does what you want it to do intuitively. If you are building an app with the state stored in plain objects, auto recursion is fine, because it is probably what you want. If you are build an app with state stored in classes, auto recursion won't recurse into class instances, which is also what you want in normal cases. So although it might be 'magic' or 'confusion' that `observable` threats simple objects differently from class instances, it is also a nice example of convention over configuration. But for me the most important reason to keep it is to avoid confusion during upgrade. If this behavior is changed, people have to check *all* their invocations of `observable(object / array)` to see whether stuff will work as expected because they might be currently relying on  auto recursion, or that they have to add (or remove) modifiers. So I think it would result in a really bad upgrade experience.
* 4. I'll deprecate and mark it for deletion in 2.1 as well. Can always decide to keep it if really needed. I think it is an anti pattern and hopefully 2.0 has all the power to solve situations more elegantly than `untracked` does. If necessarily by even creating your own `Atom`s or `Reaction`s.

**Comment by mweststrate** [comment]
MobX2 has been released with the above changes.
